### PR TITLE
2213 custom connector front end

### LINKED ISSUE #2213: Add functionality to upload connector templates
### Overview
We want to allow QA to test connectors from the connector builder and allow customers to create and upload their own connectors. 

A very rough proof-of-concept can be found at `origin/custom_connector_templates`, this was just meant as an experiment but when this feature came up in planning it made sense to share it, even in its incomplete state.

### Code Changes
- Modified the connector registry to store the actual configs, datasets, and other assets instead of just file paths. This was to standardize the contents of the registry and allow us to also read configs, and datasets from in the newly created `custom_connector_template` database table.
- Added the `/connector_template/register` endpoint to be able to upload a SaaS connector template (a zip file containing a config and dataset).
  - The parts of the connector template are stored to the database and added to the registry so that a connector can be instantiated from this newly registered template
- The SaaS connector template validation that occurs at server startup now also loads custom templates from the database
- Custom functions are now loaded during server startup and dynamically when a template is registered via the API.

### Concerns
I (@galvana), stopped working on this POC once I started having concerns about keeping these templates up-to-date. A template is validated during registration so we can assert that it will work in the current version of Fides but this cannot be guaranteed as we develop the SaaS framework or refactor Fides in general
- the SaaS config spec can change
- the modules referenced in a custom override function can change location or functionality
- we might need to change the custom override interface

### Brainstorming Ideas
I started thinking it might be best to keep custom connector templates in separate repos and structure those projects in a way that we can run continuous tests against new versions of Fides. A Fides instance could then reference these well-structured connector repos and import the templates that way.

We might also decide that we want to solve this challenge in a future iteration, any thoughts @NevilleS?

### PR DESCRIPTION
**Pull request by galvana**
Closes #2213

### Code Changes

* [x] Added `react-dropzone` dependency
* [x] Added `ConnectorTemplateUploadModal`
* [x] Added `connector-template.slice.ts` to make the call to `/connector_template/register`
* [x] Added `Upload template` button in the choose connection screen (only accessible to users with the `CONNECTOR_REGISTER_TEMPLATE` scope)

### Steps to Confirm

* [x] Upload a zip file via the UI ([planet_express.zip](https://github.com/ethyca/fides/files/11163321/planet_express.zip))

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### Description Of Changes

Added a button to the Choose connection screen which opens a modal to upload a connector template. This button only shows up for users with the `CONNECTOR_REGISTER_TEMPLATE` scope.

### CODE REVIEW

**Review by allisonking — APPROVED** [review]
awesome job @galvana !! I've got mostly small comments but nothing blocking. 

I did notice I could upload the planet express zip twice and I got a success message. Planet express still only shows up once though, so maybe that's okay?

File: clients/admin-ui/src/features/connector-templates/ConnectorTemplateUploadModal.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  ModalHeader,
+  ModalOverlay,
+  Text,
+  useToast,
+} from "@fidesui/react";
+import { FetchBaseQueryError } from "@reduxjs/toolkit/dist/query/fetchBaseQuery";
+import React, { useState } from "react";
+import { useDropzone } from "react-dropzone";
+
+import { getErrorMessage } from "../common/helpers";
```
[the remark this answers is not in the document]
Comment (galvana) [reply clients/admin-ui/src/features/connector-templates/ConnectorTemplateUploadModal.tsx:18]: Updated to this
```
import { getErrorMessage } from "~/features/common/helpers";
import { errorToastParams, successToastParams } from "~/features/common/toast";
import { useRegisterConnectorTemplateMutation } from "~/features/connector-templates/connector-template.slice";
```

File: clients/admin-ui/src/features/connector-templates/connector-template.slice.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+export const connectorTemplateSlice = createSlice({
+  name: "connectorTemplate",
+  initialState,
+  reducers: {
+    setLoading: (draftState, action: PayloadAction<boolean>) => {
+      draftState.loading = action.payload;
+    },
+    setError: (draftState, action: PayloadAction<string | null>) => {
+      draftState.error = action.payload;
+    },
```
[the remark this answers is not in the document]
Comment (galvana) [reply clients/admin-ui/src/features/connector-templates/connector-template.slice.ts:22]: Makes sense, I cleaned this up and removed `types.ts` where I had a custom state type defined.

File: clients/admin-ui/src/features/datastore-connections/add-connection/ChooseConnection.tsx [hunk 3]
```diff
@@ -15,11 +16,20 @@ import {
...
+  useRef,
+  useState,
+} from "react";
 import { useDispatch } from "react-redux";
 
 import { useAppSelector } from "~/app/hooks";
+import Restrict from "~/features/common/Restrict";
+import { ScopeRegistryEnum } from "~/types/api";
 
+import ConnectorTemplateUploadModal from "../../connector-templates/ConnectorTemplateUploadModal";
```
[the remark this answers is not in the document]
Comment (galvana) [reply clients/admin-ui/src/features/datastore-connections/add-connection/ChooseConnection.tsx:32]: Updated!

File: clients/admin-ui/src/features/datastore-connections/add-connection/ChooseConnection.tsx [hunk 4]
```diff
@@ -95,7 +110,23 @@ const ChooseConnection: React.FC = () => {
             type="search"
           />
         </InputGroup>
+        <Restrict scopes={[ScopeRegistryEnum.CONNECTOR_TEMPLATE_REGISTER]}>
+          <Button
+            colorScheme="primary"
+            type="submit"
+            width="20%"
```
[the remark this answers is not in the document]
Comment (galvana) [reply clients/admin-ui/src/features/datastore-connections/add-connection/ChooseConnection.tsx:117]: Ooh! That's nice

**Comment by galvana** [comment]
Thanks for the review @allisonking! I've made the requested changes. Also, the upload is working as expected, since it's the same template it just updates the existing one 👍

**Review by allisonking — APPROVED** [review]
looks great!! 💥

**Pull request merged**